subprojects/logging/src/main/java/org/gradle/internal/featurelifecycle/LoggingDeprecatedFeatureHandler.java

```diff
@@ -45,36 +47,39 @@ public class LoggingDeprecatedFeatureHandler implements FeatureHandler<Deprecate
 
     private final Set<String> messages = new HashSet<String>();
     private boolean deprecationsFound = false;
-    private UsageLocationReporter locationReporter;
+    private ProblemDiagnosticsFactory problemDiagnosticsFactory = new NoOpProblemDiagnosticsFactory();
 
     private WarningMode warningMode = WarningMode.Summary;
     private BuildOperationProgressEventEmitter progressEventEmitter;
     private GradleException error;
 
-    public LoggingDeprecatedFeatureHandler() {
-        this.locationReporter = DoNothingReporter.INSTANCE;
-    }
-
-    public void init(UsageLocationReporter reporter, WarningMode warningMode, BuildOperationProgressEventEmitter progressEventEmitter) {
-        this.locationReporter = reporter;
+    public void init(ProblemDiagnosticsFactory problemDiagnosticsFactory, WarningMode warningMode, BuildOperationProgressEventEmitter progressEventEmitter) {
+        this.problemDiagnosticsFactory = problemDiagnosticsFactory;
         this.warningMode = warningMode;
         this.progressEventEmitter = progressEventEmitter;
     }
 
     @Override
-    public void featureUsed(DeprecatedFeatureUsage usage) {
+    public void featureUsed(final DeprecatedFeatureUsage usage) {
         deprecationsFound = true;
+        ProblemDiagnostics diagnostics = problemDiagnosticsFactory.forCurrentCaller(new ProblemDiagnosticsFactory.StackTraceTransformer() {
+            @Override
+            public List<StackTraceElement> transform(StackTraceElement[] original) {
+                return FeatureUsage.calculateStack(usage.getCalledFrom(), original);
```

Comment:
:thinking: Should this be an actual class where the `calculateStack()` method is as well? The message seems somewhat orphaned on `FeatureUsage` now.

Reply:
This is in a separate class now.